pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,520 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math"
+	"net/http"
+	"os"
+	"reflect"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	cmg "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	mgr       MetricsManager
+	srv       *http.Server
+	wg        *sync.WaitGroup
+	statusMap map[int]string = map[int]string{
+		0: "Success",
+		1: "Failure",
+		2: "Unknown",
+	}
+)
+
+const (
+	httpPattern = "/metrics"
+	addr        = "localhost:12134"
+	httpAddr    = "http://" + addr + httpPattern
+)
+
+type fakeOpStatus struct {
+	statusCode int
+}
+
+func (s *fakeOpStatus) String() string {
+	if str, ok := statusMap[s.statusCode]; ok {
+		return str
+	}
+	return "Unknown"
+}
+
+func initMgr() {
+	wg = &sync.WaitGroup{}
+	wg.Add(1)
+	mgr = NewMetricsManager()
+	var err error
+	if srv, err = mgr.StartServing(httpPattern, addr, nil, wg); err != nil {
+		log.Fatalf("failed to start serving [%v]", err)
+	}
+}
+
+func shutdown() {
+	if err := srv.Shutdown(context.Background()); err != nil {
+		panic(err)
+	}
+	wg.Wait()
+}
+
+func TestMain(m *testing.M) {
+	initMgr()
```

Comment:
This is on purposely set up to avoid potential "port in use" errors. After a HTTP server has been shutdown, there is a short period of time for the OS to mark the port available, trying to bind to that port would cause failure.
"That will cause problems & flaky if there is any state that changes across test." could you be a bit more specific?

Reply:
Global state across tests means that one test can change things so that another test fails. But that failing test will succeed when run in isolation, making debugging very hard. That's the main reason in my experience.

The second reason is that it makes it impossible to parallelize the tests, although that's more of a theoretical issue now.

To avoid the port-in-use errors pass a 0 port, which will cause a port to be auto-allocated. The address can be recovered from svr.Addr. 

Doing this is a good idea anyway in case there are other tests or things running on the system that are allocating a port; this should be changed in csi-lib-utils as well which is yet another argument for your comment above :)

Reply:
My apologies I have not made myself clear.
While I complete agree with your points on global state could end up cause, but it's a bit special here for http Servers without customized Handler. I have made the code change to adapt the comments.
If http.Handle(....) is used to register request handlers and no handler is supplied to create a [Server](https://golang.org/pkg/net/http/?m=all#Server), the DefaultServeMux will be used which itself is global.